### PR TITLE
Equals hashcode fixing

### PR DESCRIPTION
**Pull request by JozeFe**
I refactor methods equals and hashcode,
-use business key instead of Id
-use instanceof instead of getClass()
-use getters on "other object" instead of taking advantages of visibility of other object's private field that IDE automatically generate.
-add missing no-arg constructor in some entities

### CODE REVIEW

File: fleet-management/src/main/java/cz/fi/muni/pa165/entity/Employee.java [hunk 1]
```diff
@@ -19,6 +19,16 @@
     @Column(nullable=false)
     private String surname;
 
+    /**
+     * Initialize new Employee object. For serialization uses.
+     */
+    public Employee() {
```
Comment (trebi) [inline fleet-management/src/main/java/cz/fi/muni/pa165/entity/Employee.java]: What does it mean "for serialization uses"? Where it is actually used?

File: fleet-management/src/main/java/cz/fi/muni/pa165/entity/InspectionInterval.java [hunk 2]
```diff
@@ -60,17 +61,20 @@ public void setDays(int days) {
...
+        if (days != that.getDays()) return false;
+        return name != null ? name.equals(that.getName()) : that.getName() == null;
 
     }
 
     @Override
     public int hashCode() {
-        return id != null ? id.hashCode() : 0;
+        int result = name != null ? name.hashCode() : 0;
+        result = 31 * result + days;
```
Comment (trebi) [inline fleet-management/src/main/java/cz/fi/muni/pa165/entity/InspectionInterval.java]: Why is it multiplied by 31?

File: fleet-management/src/main/java/cz/fi/muni/pa165/entity/Employee.java (hunk 1 above)
Comment (JozeFe) [inline fleet-management/src/main/java/cz/fi/muni/pa165/entity/Employee.java]: I just copy paste existing comment in one entity, I can remove it, but hibernate need no-arg constructor in entity ...

File: fleet-management/src/main/java/cz/fi/muni/pa165/entity/InspectionInterval.java (hunk 2 above)
Comment (JozeFe) [inline fleet-management/src/main/java/cz/fi/muni/pa165/entity/InspectionInterval.java]: Auto generate hashCode from idea, is fine, it just generate a "unique" number for specific entity, necessary in some collections.

File: fleet-management/src/main/java/cz/fi/muni/pa165/entity/Employee.java (hunk 1 above)
Comment (trebi) [inline fleet-management/src/main/java/cz/fi/muni/pa165/entity/Employee.java]: @JozeFe so make it `protected` please

**Comment by trebi** [comment]
I still thinks that `equals()` is not implemented correctly. I would go for the implementation like this (example is in C# taken from [this blog post](http://enterprisecraftsmanship.com/2014/11/08/domain-object-base-class/)):
```
 public override bool Equals(object obj)
    {
        var compareTo = obj as Entity;
 
        if (ReferenceEquals(compareTo, null))
            return false;
 
        if (ReferenceEquals(this, compareTo))
            return true;
 
        if (GetRealType() != compareTo.GetRealType())
            return false;
 
        if (!IsTransient() && !compareTo.IsTransient() && Id == compareTo.Id)
            return true;
 
        return false;
    }
```

**Comment by trebi** [comment]
@JozeFe you should not merge a request if there are still some unresolved issues :-(

**Comment by JozeFe** [comment]
@trebi I'm sorry for that, my mistake. I somehow miss last comment :( . But I think now it's right implemented anyway. Here I post something from Effective Java 2nd Edition (Item 8):

The equals method implements an equivalence relation. It is:
• **Reflexive**: For any non-null reference value x , x.equals(x) must return true .
• **Symmetric**: For any non-null reference values x and y , x.equals(y) must re-
turn true if and only if y.equals(x) returns true .
• **Transitive**: For any non-null reference values x , y , z , if x.equals(y) returns
true and y.equals(z) returns true , then x.equals(z) must return true .
• **Consistent**: For any non-null reference values x and y , multiple invocations
of x.equals(y) consistently return true or consistently return false , pro-
vided no information used in equals comparisons on the objects is modified.
• For any **non-null** reference value x , x.equals(null) must return **false** . 

I can add tests for that relations, if it's necessary.
End summary in "Item 8: Obey the general contract when overriding equals" from Effective Java

Putting it all together, here’s a recipe for a high-quality equals method:
1. Use the == operator to check if the argument is a reference to this object.
2. Use the instanceof operator to check if the argument has the correct type.
3. Cast the argument to the correct type. Because this cast was preceded by an
instanceof test, it is guaranteed to succeed.
4. For each “significant” field in the class, check if that field of the argument
matches the corresponding field of this object. 
5. When you are finished writing your equals method, ask yourself three
questions: Is it symmetric? Is it transitive? Is it consistent? 

Here are a few final caveats:
• Always override hashCode when you override equals (Item 9).
• Don’t try to be too clever. If you simply test fields for equality, it’s not hard
to adhere to the equals contract. If you are overly aggressive in searching for
equivalence, it’s easy to get into trouble. It is generally a bad idea to take any
form of aliasing into account. For example, the File class shouldn’t attempt to
equate symbolic links referring to the same file. Thankfully, it doesn’t.
• Don’t substitute another type for Object in the equals declaration.

And I think the equals methods are correct right now. If not I will correct my mistake and avoid another merge if there is unresolved issues. :)

**Comment by trebi** [comment]
@JozeFe no problem, it happens. Thanks for a nice reference to Effective Java. However, I think that `equals()` is not correct yet anyway. For example:
- for the `Employee` - `equals()`returns `true`, if the employees have the same name and surname. It's technically possible, because neither `name` nor `surname` are unique, and the situation could easily occur in reality (e.g. father and son quite often have the same names).
- for the `Inspection` entity - `equals()`returns `true` for all inspections that occurred in the same day! This is without any doubt wrong!
- for the `Journey` entity - you compare `distance`, `borrowedAt` and `returnedAt` attributes. However, `distance` `returnedAt` could be null. So again, all journeys that began in the same day and did not finished yet are considered equal.

**Comment by JozeFe** [comment]
@trebi And that's right, that is exactly how it should work :) . If you now about two persons only name and surname and they are the same it should return true :). It's correct implementation.
From your post there are Reference equality (with "=="), Identifier equality (by id), Logical equality (our implementation). 
In lectures and Effective Java were said we shouldn't use Id that mean we should use logical implementation. So 99% of equals implementation are logical. In some extremly specific projects you may want to use Identifier equality but no need in ours. :) 
Two different rows in the database can actually be logical equal. That's not wrong implementation of equal method. It's most common implementation and correct.

**Comment by trebi** [comment]
> When you create an object with new(), and then save it into a session, Hibernate now knows that whenever you query for an object and find that particular object, Hibernate should return you that instance of the object.

Consider
```
journey1 = new Journey(vehicle1);
journey2 = new Journey(vehicle2);
employee1.addJourney(journey1);
employee2.addJourney(journey2);
```

It will behave like this:
```
employee1.getJourneys()[0].equals(
    employee2.getJourneys()[0]
); // TRUE

employee1.getJourneys()[0].getVehicle().equals(
    employee2.getJourneys()[0].getVehicle()
); // FALSE
```
But why should it?
See commit: https://github.com/Qerts/pa165-java/commit/c096406b73888bf6d7ce4fe367578535aeeddfef

Isn't it quite odd?

**Pull request merged**